File: 3Sum/3Sum.cpp
class Solution {
public:

    void Sort(std::vector<int>& in){
        for(unsigned int i = 1; i < in.size(); ++i){
            for(unsigned int j = 0; j < i; ++j){
                if(in[j] > in[i]){
                    int tmp = in[j];
                    in[j] = in[i];
                    in[i] = tmp;
                }
            }
        }
    }

    vector<vector<int> > threeSum(vector<int> &num) {
        std::vector<std::vector<int> >  res;
        
        if(num.size() < 3){
            return res;
        }
        
        Sort(num);
        
        for(unsigned int i = 0; i < num.size() - 2; ++i){
            
            int start = i + 1;
            int end = num.size() - 1;
            int target = -num[i];
            while(start < end){
                int sum = num[start] + num[end];
                if(sum == target){
                    std::vector<int>    element;
                    std::vector<int>    prev;
                    
                    if(!res.empty()){
                        prev = res[res.size() - 1];
                        if((prev[0] == num[i]) && (prev[1] == num[start]) && (prev[2] == num[end])){
                            ++start;
                            --end;
                            continue;
                        }
                    }
                    
                    element.push_back(num[i]);
                    element.push_back(num[start]);
                    element.push_back(num[end]);
                    res.push_back(element);
                    
                    ++start;
                    --end;
                    
                }
                else if(sum < target){
                    ++start;
                }
                else{
                    --end;
                }
            }
            // Weird
            // This exception could be implicitly handled inside, but it would not pass the test
            // Explicit handling could save time
            while(num[i] == num[i + 1]){
                ++i;
            }
        }
        return res;
    }
};

File: 3Sum/3Sum.java
public class Solution {
    
    
    
    public List<List<Integer>> threeSum(int[] num) {
        
        if(num.length < 3){
            return new ArrayList<List<Integer>>();
        }
        
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        Arrays.sort(num);
        
        for(int i = 0; i < num.length - 2; ++i){
            int a = num[i];
            int sum = -a;
            int j = i + 1;
            int k = num.length - 1;
            while(j < k){
                int b = num[j];
                int c = num[k];
                if(b + c < sum){
                    ++j;
                }
                else if(b + c > sum){
                    --k;
                }
                else{
                    List<Integer> newList = new ArrayList<Integer>();
                    newList.add(a);
                    newList.add(b);
                    newList.add(c);
                    ret.add(newList);
                    do{
                        ++j;
                        
                    }while(j < num.length && j < k && num[j] == num[j - 1]);
                    do{
                        --k;
                    }while(k > 0 && k > j && num[k] == num[k + 1]);
                }
            }
            while(i + 1 < num.length - 2 && num[i] == num[i + 1]){
                ++i;
            }
        }
        
        return ret;
    }
}
